src/inference/ess.jl

```diff
@@ -0,0 +1,132 @@
+"""
+    ESS
+
+Elliptical slice sampling algorithm.
+
+# Examples
+```jldoctest; setup = :(Random.seed!(1))
+julia> @model gdemo(x) = begin
+           m ~ Normal()
+           x ~ Normal(m, 0.5)
+       end
+gdemo (generic function with 2 methods)
+
+julia> sample(gdemo(1.0), ESS(), 1_000) |> mean
+Mean
+
+│ Row │ parameters │ mean     │
+│     │ Symbol     │ Float64  │
+├─────┼────────────┼──────────┤
+│ 1   │ m          │ 0.811555 │
+```
+"""
+struct ESS{space} <: InferenceAlgorithm end
+
+ESS() = ESS{()}()
+ESS(space::Symbol) = ESS{(space,)}()
+
+mutable struct ESSState{V<:VarInfo} <: AbstractSamplerState
+    vi::V
+end
+
+ESSState(model::Model) = ESSState(VarInfo(model))
+
+function Sampler(alg::ESS, model::Model, s::Selector)
+    # sanity check
+    space = getspace(alg)
+    if isempty(space)
+        pvars = get_pvars(model)
+        length(pvars) == 1 ||
+            error("[ESS] no symbol specified to sampler although there is not exactly one model parameter ($pvars)")
+    end
+
+    state = ESSState(model)
+    info = Dict{Symbol, Any}()
+
+    return Sampler(alg, info, s, state)
+end
+
+# always accept in the first step
+function step!(::AbstractRNG, ::Model, spl::Sampler{<:ESS}, ::Integer; kwargs...)
+    return Transition(spl)
+end
+
+function step!(
+    rng::AbstractRNG,
+    model::Model,
+    spl::Sampler{<:ESS},
+    ::Integer,
+    ::Transition;
+    kwargs...
+)
+    # recompute log-likelihood in logp
+    vi = spl.state.vi
+    if spl.selector.tag !== :default
+        runmodel!(model, vi, spl)
+    end
+
+    # obtain previous sample
+    f = copy(vi[spl])
+
+    # sample log-likelihood threshold for the next sample
+    threshold = getlogp(vi) - randexp(rng)
+
+    # sample from the prior
+    runmodel!(model, vi, spl)
+    ν = copy(vi[spl])
+
+    # sample initial angle
+    θ = 2 * π * rand(rng)
+    θₘᵢₙ = θ - 2 * π
+    θₘₐₓ = θ
+
+    while true
+        # compute proposal
+        sinθ, cosθ = sincos(θ)
+        @. vi[spl] = f * cosθ + ν * sinθ
+
+        # recompute log-likelihood and check if threshold is reached
+        resetlogp!(vi)
+        model(vi, spl)
```

Comment:
Is this the correct way to just recompute the log-likelihood for the updated sample? Is `resetlogp!` needed?

Reply:
If you consolidate these to `runmodel!` you don't need `resetlogp!`. You're essentially doing all of that function in two lines right here, it would be slightly more idiomatic to use `runmodel!`.

Reply:
OK, thanks!